csrc/val_graph.cpp

```diff
@@ -0,0 +1,463 @@
+  NVF_ERROR(
+      unique_definitions_.emplace(val_disjoint_set, def_groups).second,
+      "Multiple defining groups for ",
+      nvfuser::toString(val_disjoint_set));
+
+  ExprGroups use_groups;
+  for (auto use : uses) {
+    const ExprGroup& expr_set =
+        disjointExprSets().initializeSet(use).first->second;
+    use_groups.pushBack(expr_set);
+  }
+  // TODO-NM: use_groups can be empty. Should it be still mapped?
+  NVF_ERROR(
+      unique_uses_.emplace(val_disjoint_set, use_groups).second,
+      "Multiple use groups for ",
+      nvfuser::toString(val_disjoint_set));
+}
+
+void ValGraph::initializeVal(Val* val) {
+  VectorOfUniqueEntries<Expr*> defs;
+  if (val->definition()) {
+    defs.pushBack(val->definition());
+  }
+  VectorOfUniqueEntries<Expr*> uses;
+  for (Expr* use : val->uses()) {
+    uses.pushBack(use);
+  }
+  initializeVal(val, defs, uses);
+}
+
+bool ValGraph::exprsMap(Expr* first, Expr* second, bool forward) const {
+  NVF_ERROR(first);
+  NVF_ERROR(second);
+
+  if (!first->sameOp(second)) {
+    return false;
+  }
+
+  std::vector<Val*> first_vals = forward ? first->inputs() : first->outputs();
+  std::vector<Val*> second_vals =
+      forward ? second->inputs() : second->outputs();
+
+  NVF_ERROR(
+      first_vals.size() == second_vals.size(),
+      "Expected number of ",
+      (forward ? "inputs" : "outputs"),
+      " to match for\n",
+      first->toString(),
+      second->toString());
+
+  for (const auto i : c10::irange(first_vals.size())) {
+    if (!disjointValSets().permissiveAreMapped(
+            first_vals.at(i), second_vals.at(i))) {
+      return false;
+    }
+  }
+
+  // Special handling for backprop of merge
+  if (first->isA<Merge>() && !forward) {
+    // Can't back prop through merge without making sure one input actually
+    // matches. This can be done on a map or extent basis.
+    auto merge0 = first->as<Merge>();
+    auto merge1 = second->as<Merge>();
+
+    auto extent_0o = merge0->outer()->extent();
+    auto extent_0i = merge0->inner()->extent();
+    auto extent_1o = merge1->outer()->extent();
+    auto extent_1i = merge1->inner()->extent();
+
+    auto extent_o_match = extent_0o->sameAs(extent_1o) ||
+        (extent_0o->isConstInt() && extent_1o->isConstInt() &&
+         extent_0o->evaluate() == extent_1o->evaluate()) ||
+        disjointValSets().permissiveAreMapped(merge0->outer(), merge1->outer());
+
+    auto extent_i_match = extent_0i->sameAs(extent_1i) ||
+        (extent_0i->isConstInt() && extent_1i->isConstInt() &&
+         extent_0i->evaluate() == extent_1i->evaluate()) ||
+        disjointValSets().permissiveAreMapped(merge0->inner(), merge1->inner());
+
+    if (!(extent_o_match || extent_i_match)) {
+      return false;
+    }
+  }
+
+  // TODO: For now we're using same as, however we could know what val's are
+  // exactly the same given the exact map. We might want to pipe that
+  // information through to here.
```

Comment:
I don't understand this sentence. Is this saying that if we already have an exact map, then we will know what is equivalent? But aren't we building the exact map, which means that we never have an existing an exact map?

Reply:
I think what it meant was it would be possible to use the exact graph to help discover equality rather than just using `sameAs`.

Reply:
While I agree with the description that exact graph can be used for discover equality, I felt strange that this is marked as a TODO instead of a note. I thought at some point, there was a code using `sameAs`, and you left a comment saying that we should not use sameAs here, we should instead use the exact graph. But maybe you didn't include this piece of code in this PR.

Reply:
I'm not 100% sure as this comment is from Christian's original PR.